File: 1920-buildArrayFromPermutation/buildArray.hs
buildArray :: [Int] -> [Int]
buildArray nums = [nums !! x | x <- nums]

buildArrayDesugarized nums = do
  x <- nums
  return $ nums !! x

buildArrayFullyDesugarized :: [Int] -> [Int]
buildArrayFullyDesugarized nums = nums >>= \x -> return $ nums !! x

File: 1920-buildArrayFromPermutation/buildArray.c
#include <stdlib.h>
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int *buildArray(int *nums, int numsSize, int *returnSize) {
  *returnSize = numsSize;
  int *ret = malloc(numsSize * sizeof(int));
  for (int i = 0; i < numsSize; i++) {
    ret[i] = nums[nums[i]];
  }
  return ret;
}
